### PR TITLE
Add `Module::version`

### PR DESCRIPTION
**Pull request by klimov-paul**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | yes |
| Tests pass? | yes |
| Fixed issues | #12726 |

### CODE REVIEW

File: framework/base/Module.php [hunk 1]
```diff
@@ -283,6 +300,58 @@ public function setLayoutPath($path)
...
+     *
+     * ```php
+     * function (Module $module) {
+     *     //return string|integer
+     * }
+     * ```
+     *
+     * @since 2.0.11
+     */
+    public function setVersion($version)
```
Comment (samdark) [inline framework/base/Module.php]: Why do we need a setter?

File: framework/base/Module.php [hunk 2]
```diff
@@ -283,6 +300,58 @@ public function setLayoutPath($path)
...
+
+    /**
+     * Returns default module [[version]] based last modification time of the [[basePath]].
+     * Child class may override this method to provide more specific version detection.
+     * @return string|integer module version.
+     * @since 2.0.11
+     */
+    protected function defaultVersion()
+    {
+        $versionTimestamp = @filemtime($this->getBasePath());
```
Comment (samdark) [inline framework/base/Module.php]: It would return local file date instead of date of a file from repository, right?

**Review by samdark — COMMENTED** [review]
It's it better to introduce real version reading or a meta file?

File: framework/base/Module.php (hunk 1 above)
Comment (klimov-paul) [inline framework/base/Module.php]: It allows you to set version as a PHP callback, which will return the actual value intead of extending class. It is crucial for `Application`, allowing specification of version logic in config file:

```php
return [
    'name' => 'MyApplication',
    'basePath' => dirname(__DIR__),
    'version' => function ($application) {
        return exec("cd {$application->basePath}; git rev-parse HEAD");
    },
    // ...
];
```

File: framework/base/Module.php (hunk 2 above)
Comment (klimov-paul) [inline framework/base/Module.php]: Although there is a problem here: current implementation will pick up last modification date from `runtime` directory as well, which will cause it to always reference current date time, as `runtime` is modified almost at any program run.

Comment (rob006) [inline framework/base/Module.php]: > current implementation will pick up last modification date from runtime directory as well

Did you test it? Modifications in `runtime` should not change mtime of root directory.

Comment (klimov-paul) [inline framework/base/Module.php]: Improved to use VCS files as a timestamp source.

**Comment by klimov-paul** [comment]
Just noticed: `yii\base\Application` already has a `version` public property:
https://github.com/yiisoft/yii2/blob/master/framework/base/Application.php#L100

**Comment by klimov-paul** [comment]
Legacy `Application::version` removed.
Thus this PR produces a slight BC break, as it will return version based on timestamp instead of `1.0` be default.

**Comment by rob006** [comment]
Timestamp-based versions are useless anyway. mtime is completely unrelated to real application version - changes in application not always change mtime of project root.

**Comment by klimov-paul** [comment]
> Timestamp-based versions are useless anyway. 

Make sense. I have changed default value to `1.0` keeping the BC.

**Review by samdark — APPROVED** [review]

File: framework/base/Module.php (hunk 2 above)
Comment (dynasource) [inline framework/base/Module.php]: there shouldnt be any calculated defaultVersion. Thats to arbitrary. Its either the user who is defining it or no version.

**Comment by dynasource** [comment]
> Just noticed: yii\base\Application already has a version public property

This shows that its an idea to open a topic to further cleanup Application & Module. Bootstrap for instance, is something that is useful too for Modules.

**Review by SilverFire — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Pull request merged**